corehq/apps/registry/models.py

```diff
@@ -0,0 +1,127 @@
+            query = query.filter(is_active=is_active)
+        return query
+
+    def accessible_to_domain(self, domain, slug=None, has_grants=False):
+        """
+        :param domain: Domain to get registries for
+        :param slug: (optional) Filter registries by slug
+        :param has_grants: (optional) Set to 'True' to only include registries for which the domain has grants
+        """
+        query = (
+            self.filter(is_active=True)
+            .filter(
+                invitations__domain=domain,
+                invitations__accepted_on__isnull=False,
+                invitations__rejected_on__isnull=True
+            )
+        )
+        if slug:
+            query = query.filter(slug=slug)
+        if has_grants:
+            query = query.filter(grants__to_domains__contains=[domain])
+        return query
+
+
+class DataRegistry(models.Model):
+    domain = models.CharField(max_length=255)
+    name = models.CharField(max_length=255)
+    # slug used for referencing the registry in app suite files, APIs etc.
+    slug = AutoSlugField(populate_from='name', unique_with='domain', slugify=slugify_remove_stops)
+    description = models.TextField(blank=True)
+    is_active = models.BooleanField(default=True)
+    schema = JSONField(null=True, blank=True)
+
+    created_on = models.DateTimeField(auto_now_add=True)
+    modified_on = models.DateTimeField(auto_now=True)
+
+    objects = RegistryManager()
+
+    class Meta:
+        unique_together = ('domain', 'slug')
+
+    def deactivate(self):
+        self.is_active = False
+        self.save()
+
+    def get_granted_domains(self, domain):
+        self.check_access(domain)
+        return set(
+            self.grants.filter(to_domains__contains=[domain])
+            .values_list('from_domain', flat=True)
+        )
+
+    def check_access(self, domain):
+        if not self.is_active:
+            raise RegistryAccessDenied()
+        invites = self.invitations.filter(domain=domain)
+        if not invites:
+            raise RegistryAccessDenied()
+        invite = invites[0]
+        if not invite.accepted_on or invite.rejected_on:
+            raise RegistryAccessDenied()
+        return True
+
+
+class RegistryInvitation(models.Model):
+    registry = models.ForeignKey("DataRegistry", related_name="invitations", on_delete=models.CASCADE)
+    domain = models.CharField(max_length=255)
+    created_on = models.DateTimeField(auto_now_add=True)
+    accepted_on = models.DateTimeField(null=True, blank=True)
+    accepted_by = models.ForeignKey(
+        User, related_name="registry_accepted_invitations", on_delete=models.CASCADE, null=True, blank=True
+    )
+    rejected_on = models.DateTimeField(null=True, blank=True)
+    rejected_by = models.ForeignKey(
+        User, related_name="registry_rejected_invitations", on_delete=models.CASCADE, null=True, blank=True
+    )
+
+    class Meta:
+        unique_together = ("registry", "domain")
```

Comment:
Good question. They way I've been thinking about this is that a domain can change their mind at any point. If they accept and then reject then both sets of fields would be filled in. If they change their mind after rejecting then we'd need to blank out the reject fields. 

The 'owing' domain could re-invite them which would blank out both accept and reject.

I did think about creating multiple invitations per domain but that makes it harder to check a domains status since you'd need to find the most recent invitation. Enforcing only one makes that simpler.

Another possibility is that we could move some of this data to an audit log model which would keep track of historical invitation changes and then this model would just have a 'status' field and a 'date_modified' field.

Reply:
Yeah, I like the idea of a separate audit log to keep this from getting too cluttered.  Though it might be a YAGNI.  Depends on 1) how much we care about having a precise record and 2) how often invitations actually see churn in practice vs being set once and forgotten about.